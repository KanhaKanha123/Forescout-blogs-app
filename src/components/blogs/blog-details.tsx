import { FC, Fragment, ReactElement, useState } from "react";
import { Link, useParams } from "react-router-dom";
import {
    Button,
    Card,
    HandleLoadingEmptyErrorState,
    PaginatedItems
} from "../shared-comonents";
import {
    BlogsWrapper,
    AuthourText,
    CommentsWrapper,
    CommentsTextAndButtonWrapper,
    Text,
    Header,
    ReactionsContainer,
    BodyContainer,
    TitleTextWrapper,
    BlogBody,
    AddNewCommentsWrapper
} from "./style";
import {
    useGetSingleBlog,
    useGetUserData,
} from "../../custom-hooks";
import { useAppContext } from "../../store/context";
import config from "../../config";
import { BlogComments } from "./blog-comments";
import { AddNewComments } from "./add-new-comments";

/**
 * This component is responsible to render the blog details and the comment list component
 * @returns ReactElement
 */
export const BlogDetails: FC = (): ReactElement => {

    const [isAddNewCommentFormShow, setIsAddNewCommentFormShow] = useState<boolean>(false);

    //get blogId from url params
    const { blogId } = useParams<string>();

    //get  blog data from hook
    const { blog, isLoading, error } = useGetSingleBlog(blogId!);

    //get userData data from hook
    const { userData } = useGetUserData(blog?.userId!);

    //get totalCountComments data from context provider
    const { totalCountComments } = useAppContext();

    const showHideAddNewCommentBoxHandler = () => setIsAddNewCommentFormShow(prevState => !prevState);

    return (
        <HandleLoadingEmptyErrorState data={blog} isLoading={isLoading} error={error}>
            <BlogsWrapper aria-label="container with blog details">
                <Card
                    width="1000px"
                    height="auto"
                    padding="15px"
                    backgroundcolor="#f4f4f4">
                    <Header>
                        <TitleTextWrapper aria-label="blog title here">
                            {blog?.title}
                        </TitleTextWrapper>
                        <Link aria-label="link to go back" className="linkStyle" to="/">{`<- Go Back`}</Link>
                    </Header>
                    <BodyContainer aria-label="authour name and blog description container">
                        <AuthourText aria-label="authour name">Authour: {userData?.firstName} {userData?.lastName}</AuthourText>
                        <BlogBody aria-label="authour name">{blog?.body}</BlogBody>
                    </BodyContainer>
                    <CommentsTextAndButtonWrapper aria-label="comments static text and add new button container">
                        <ReactionsContainer>
                            <Text aria-label="comments static text">
                                Comments
                            </Text>
                        </ReactionsContainer>
                        <Button aria-label="Add new button" backgroundcolor="#4792ef" onClick={showHideAddNewCommentBoxHandler}>Add New</Button>
                    </CommentsTextAndButtonWrapper>
                    <CommentsWrapper aria-label="Comments list and pagination container">
                        <Fragment>
                            {blog?.id && <BlogComments blogId={blog?.id}></BlogComments>}
                            {config.commentsPageLimit < totalCountComments && <PaginatedItems aria-label="Pagination container" type='commentList' itemsPerPage={config.commentsPageLimit} />}
                        </Fragment>
                    </CommentsWrapper>
                </Card>
                {isAddNewCommentFormShow && <AddNewCommentsWrapper aria-label="Add new comment container">
                    <AddNewComments blogId={blog?.id!} showHideAddNewCommentBoxHandler={showHideAddNewCommentBoxHandler}></AddNewComments>
                </AddNewCommentsWrapper>}
            </BlogsWrapper>
        </HandleLoadingEmptyErrorState>
    )
}