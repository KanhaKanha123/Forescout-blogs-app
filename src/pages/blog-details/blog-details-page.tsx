import { FC, ReactElement } from "react";
import { BlogDetails } from "../../components/";

/**
 * This component is responsible render the blog details page on router request.
 * @returns ReactElement
 */
export const BlogDetailsPage: FC = (): ReactElement => {

    return <section area-label='blog details is listed here'>
        <BlogDetails />
    </section>
};