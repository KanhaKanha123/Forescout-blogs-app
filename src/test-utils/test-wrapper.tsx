import ContextProvider from '../store/context';
import { QueryClientProvider } from "react-query";
import { FC, ReactElement } from 'react';
import { client } from '../react-query/query-client';
import { BrowserRouter as Router } from 'react-router-dom';

interface TestWrapperProps {
    children: React.ReactNode
}

/**
 * This component is responsible to wrap the tests with all the important wrappers so all the time in all tests we don't have to repeat.
 * @returns ReactElement
 */
export const TestWrapper: FC<TestWrapperProps> = ({ children }): ReactElement => (
    <QueryClientProvider client={client}>
        <ContextProvider>
            <Router>
                {children}
            </Router>
        </ContextProvider>
    </QueryClientProvider>);