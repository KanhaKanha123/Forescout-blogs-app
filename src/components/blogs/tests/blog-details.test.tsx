import { render } from '@testing-library/react';
import { BlogDetails } from '../blog-details';
import { TestWrapper } from '../../../test-utils/test-wrapper';

//mock custom hook response
jest.mock('../../../custom-hooks/use-get-single-blog', () => ({
  useGetSingleBlog: () => ({
    blog: {
      body: 'hello i am here',
      id: 1,
      reactions: 5,
      tags: ['one', 'two'],
      title: 'the best blog',
      userId: 1
    },
    isLoading: false,
    error: null
  })
}));

jest.mock('../../../custom-hooks/use-get-user-data', () => ({
  useGetUserData: () => ({
    userData: {
      id: 1,
      firstName: 'test',
      lastName: 'test name',
      age: 30,
    }
  })
}));

const rederComponent = () => render(<TestWrapper>
  <BlogDetails />
</TestWrapper>);

describe('BlogDetails component', () => {

  test('renders BlogDetails component proper', () => {

    const component = rederComponent();

    expect(component).toMatchSnapshot();
  });

  test('renders BlogDetails component should have proper props', () => {

    const component = rederComponent();

    const blogBody = component.getByText(/hello i am here/i);

    const blogTitle = component.getByText(/the best blog/i);

    const authourName = component.getByText(/test test name/i);

    expect(blogBody).toBeInTheDocument();

    expect(blogTitle).toBeInTheDocument();

    expect(authourName).toBeInTheDocument();
  });
});
