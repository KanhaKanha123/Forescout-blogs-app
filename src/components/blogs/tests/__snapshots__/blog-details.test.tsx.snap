// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogDetails component renders BlogDetails component proper 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-label="container with blog details"
        class="sc-gKkivs bgfSMQ"
      >
        <div
          backgroundcolor="#f4f4f4"
          class="sc-bdDrbm hNBArv"
          height="auto"
          padding="15px"
          width="1000px"
        >
          <div
            class="sc-iCEALv dDRKGZ"
          >
            <span
              aria-label="blog title here"
              class="sc-crPhKe QNeik"
            >
              the best blog
            </span>
            <a
              aria-label="link to go back"
              class="linkStyle"
              href="/"
            >
              &lt;- Go Back
            </a>
          </div>
          <div
            aria-label="authour name and blog description container"
            class="sc-qdPnm fmmVKh"
          >
            <span
              aria-label="authour name"
              class="sc-fFCYnF cdMsrR"
            >
              Authour: 
              test
               
              test name
            </span>
            <p
              aria-label="authour name"
              class="sc-bjLqKs hcrKFR"
            >
              hello i am here
            </p>
          </div>
          <div
            aria-label="comments static text and add new button container"
            class="sc-jrICsz hYBaNZ"
          >
            <div
              class="sc-dYlqv cvZwVn"
            >
              <span
                aria-label="comments static text"
                class="sc-folRzf jAzsXY"
              >
                Comments
              </span>
            </div>
            <div
              aria-label="Add new button"
              backgroundcolor="#4792ef"
              class="sc-gtcAbF dPqWXl"
            >
              Add New
            </div>
          </div>
          <section
            aria-label="Comments list and pagination container"
            class="sc-kEbfvU kFcdHM"
          >
            <div
              class="sc-hKVryf hYMQAi"
              color="black"
              height="20px"
            >
              No Comments available
            </div>
          </section>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-label="container with blog details"
      class="sc-gKkivs bgfSMQ"
    >
      <div
        backgroundcolor="#f4f4f4"
        class="sc-bdDrbm hNBArv"
        height="auto"
        padding="15px"
        width="1000px"
      >
        <div
          class="sc-iCEALv dDRKGZ"
        >
          <span
            aria-label="blog title here"
            class="sc-crPhKe QNeik"
          >
            the best blog
          </span>
          <a
            aria-label="link to go back"
            class="linkStyle"
            href="/"
          >
            &lt;- Go Back
          </a>
        </div>
        <div
          aria-label="authour name and blog description container"
          class="sc-qdPnm fmmVKh"
        >
          <span
            aria-label="authour name"
            class="sc-fFCYnF cdMsrR"
          >
            Authour: 
            test
             
            test name
          </span>
          <p
            aria-label="authour name"
            class="sc-bjLqKs hcrKFR"
          >
            hello i am here
          </p>
        </div>
        <div
          aria-label="comments static text and add new button container"
          class="sc-jrICsz hYBaNZ"
        >
          <div
            class="sc-dYlqv cvZwVn"
          >
            <span
              aria-label="comments static text"
              class="sc-folRzf jAzsXY"
            >
              Comments
            </span>
          </div>
          <div
            aria-label="Add new button"
            backgroundcolor="#4792ef"
            class="sc-gtcAbF dPqWXl"
          >
            Add New
          </div>
        </div>
        <section
          aria-label="Comments list and pagination container"
          class="sc-kEbfvU kFcdHM"
        >
          <div
            class="sc-hKVryf hYMQAi"
            color="black"
            height="20px"
          >
            No Comments available
          </div>
        </section>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
